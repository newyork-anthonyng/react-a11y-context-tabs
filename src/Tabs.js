import React from "react";
import PropTypes from "prop-types";
import callAll from "./callAll";

const LEFT_ARROW_KEY_CODE = 37;
const RIGHT_ARROW_KEY_CODE = 39;
const UP_ARROW_KEY_CODE = 38;
const DOWN_ARROW_KEY_CODE = 40;

class Tabs extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      selectedId: null
    };
    this.tabs = [];
    this.focusTabs = {};
  }

  getChildContext() {
    const {
      forceRenderTabPanel,
      selectedTabClassName,
      selectedTabPanelClassName,
      onSelect,
      selectedId
    } = this.props;

    return {
      selectedId: this.getSelectedId(),
      onActivate: index => {
        this.setState({ selectedId: index }, () => {
          onSelect(index);
        });
      },
      onTabMount: (index, focus) => {
        if (!this.firstIndex) {
          this.firstIndex = index;

          this.setState({
            selectedId: this.firstIndex
          });
        }

        if (this.tabs.indexOf(index) === -1) {
          this.tabs = this.tabs.concat([index]);
        }
        this.focusTabs[index] = focus;
      },
      forceRenderTabPanel,
      selectedTabClassName,
      selectedTabPanelClassName
    };
  }

  getSelectedId = () => {
    return this.props.selectedId === null
      ? this.state.selectedId
      : this.props.selectedId;
  };

  handleKeyDown = ({ keyCode }) => {
    let nextTabIndex = null;

    if (this.isPreviousTabKey(keyCode)) {
      nextTabIndex = this.getPreviousTabIndex();
    } else if (this.isNextTabKey(keyCode)) {
      nextTabIndex = this.getNextTabIndex();
    }

    if (nextTabIndex !== null) {
      const newSelectedId = this.tabs[nextTabIndex];

      this.setState({ selectedId: newSelectedId }, () => {
        if (typeof this.focusTabs[newSelectedId] === "function") {
          this.focusTabs[newSelectedId]();
        }
        this.props.onSelect(newSelectedId);
      });
    }
  };

  isPreviousTabKey = keyCode => {
    return keyCode === UP_ARROW_KEY_CODE || keyCode === LEFT_ARROW_KEY_CODE;
  };

  isNextTabKey = keyCode => {
    return keyCode === RIGHT_ARROW_KEY_CODE || keyCode === DOWN_ARROW_KEY_CODE;
  };

  getPreviousTabIndex = () => {
    let tabIndex = this.tabs.indexOf(this.getSelectedId()) - 1;
    if (tabIndex < 0) tabIndex = this.tabs.length - 1;
    return tabIndex;
  };

  getNextTabIndex = () => {
    let tabIndex = this.tabs.indexOf(this.getSelectedId()) + 1;
    if (tabIndex >= this.tabs.length) tabIndex = 0;

    return tabIndex;
  };

  render() {
    const {
      innerRef,
      children,
      forceRenderTabPanel,
      selectedTabClassName,
      selectedTabPanelClassName,
      onKeyDown,
      selectedId,
      ...rest
    } = this.props;

    return (
      <div
        ref={innerRef}
        {...rest}
        onKeyDown={callAll(onKeyDown, this.handleKeyDown)}
      >
        {children}
      </div>
    );
  }
}

Tabs.propTypes = {
  forceRenderTabPanel: PropTypes.bool,
  selectedTabClassName: PropTypes.string,
  selectedTabPanelClassName: PropTypes.string,
  innerRef: PropTypes.func,
  children: PropTypes.node,
  onKeyDown: PropTypes.func,
  onSelect: PropTypes.func,
  selectedId: PropTypes.string
};

Tabs.defaultProps = {
  forceRenderTabPanel: false,
  selectedTabClassName: "react-tabs__tab--selected",
  selectedTabPanelClassName: "react-tabs__tab-panel--selected",
  selectedId: null,
  onSelect: () => {}
};

Tabs.childContextTypes = {
  onActivate: PropTypes.func,
  selectedId: PropTypes.string,
  onTabMount: PropTypes.func,
  onTabUnmount: PropTypes.func,
  forceRenderTabPanel: PropTypes.bool,
  selectedTabClassName: PropTypes.string,
  selectedTabPanelClassName: PropTypes.string
};

export default Tabs;
export {
  LEFT_ARROW_KEY_CODE,
  RIGHT_ARROW_KEY_CODE,
  UP_ARROW_KEY_CODE,
  DOWN_ARROW_KEY_CODE
};
