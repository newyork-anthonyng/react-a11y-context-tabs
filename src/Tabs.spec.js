import React from "react";
import { mount } from "enzyme";
import toJSON from "enzyme-to-json";
import Tabs, {
  LEFT_ARROW_KEY_CODE,
  RIGHT_ARROW_KEY_CODE,
  UP_ARROW_KEY_CODE,
  DOWN_ARROW_KEY_CODE
} from "./Tabs";

const defaultProps = {
  forceRenderTabPanel: false,
  selectedTabClassName: "tab--selected",
  selectedTabPanelClassName: "tabPanel--selected",
  innerRef: () => {},
  onSelect: () => {}
};

describe("Context", () => {
  it("should expose correct context", () => {
    const wrapper = mount(
      <Tabs {...defaultProps}>
        <h1>Hello World</h1>
      </Tabs>
    );

    expect(wrapper.instance().getChildContext()).toMatchSnapshot();
  });

  it("should should update selectedId with onActivate function", () => {
    const wrapper = mount(
      <Tabs {...defaultProps}>
        <h1>Hello World</h1>
      </Tabs>
    );

    const { onActivate } = wrapper.instance().getChildContext();
    onActivate("foo");

    expect(wrapper.state()).toMatchSnapshot();
  });

  it("should update this.firstIndex only once", () => {
    const wrapper = mount(
      <Tabs {...defaultProps}>
        <h1>Hello World</h1>
      </Tabs>
    );

    const { onTabMount } = wrapper.instance().getChildContext();
    onTabMount("foo");
    onTabMount("bar");

    expect(wrapper.instance().firstIndex).toEqual("foo");
  });

  it("should append tab index only if not present", () => {
    const wrapper = mount(
      <Tabs {...defaultProps}>
        <h1>Hello World</h1>
      </Tabs>
    );

    const { onTabMount } = wrapper.instance().getChildContext();
    onTabMount("foo");
    onTabMount("bar");
    onTabMount("foo");

    expect(wrapper.instance().tabs).toMatchSnapshot();
  });
});

describe("Key down events", () => {
  function mockWrapperWithSelectedId(selectedId) {
    const wrapper = mount(
      <Tabs {...defaultProps}>
        <h1>Hello World</h1>
      </Tabs>
    );

    const { onTabMount } = wrapper.instance().getChildContext();
    onTabMount("foo");
    onTabMount("bar");
    onTabMount("baz");
    wrapper.setState({ selectedId });

    return wrapper;
  }
  it("should update correctly when pressing LEFT key on first tab", () => {
    const wrapper = mockWrapperWithSelectedId("foo");

    const divEl = wrapper.find("div");
    divEl.simulate("keydown", { keyCode: LEFT_ARROW_KEY_CODE });

    expect(wrapper.state()).toMatchSnapshot();
  });

  it("should update correctly when pressing LEFT key when not on first tab", () => {
    const wrapper = mockWrapperWithSelectedId("bar");

    const divEl = wrapper.find("div");
    divEl.simulate("keydown", { keyCode: LEFT_ARROW_KEY_CODE });

    expect(wrapper.state()).toMatchSnapshot();
  });

  it("should update correctly when pressing UP key on first tab", () => {
    const wrapper = mockWrapperWithSelectedId("foo");

    const divEl = wrapper.find("div");
    divEl.simulate("keydown", { keyCode: UP_ARROW_KEY_CODE });

    expect(wrapper.state()).toMatchSnapshot();
  });

  it("should update correctly when pressing UP key when not on first tab", () => {
    const wrapper = mockWrapperWithSelectedId("bar");

    const divEl = wrapper.find("div");
    divEl.simulate("keydown", { keyCode: UP_ARROW_KEY_CODE });

    expect(wrapper.state()).toMatchSnapshot();
  });

  it("should update correctly when pressing RIGHT key on last tab", () => {
    const wrapper = mockWrapperWithSelectedId("baz");

    const divEl = wrapper.find("div");
    divEl.simulate("keydown", { keyCode: RIGHT_ARROW_KEY_CODE });

    expect(wrapper.state()).toMatchSnapshot();
  });

  it("should update correctly when pressing RIGHT key when not on last tab", () => {
    const wrapper = mockWrapperWithSelectedId("bar");

    const divEl = wrapper.find("div");
    divEl.simulate("keydown", { keyCode: RIGHT_ARROW_KEY_CODE });

    expect(wrapper.state()).toMatchSnapshot();
  });

  it("should update correctly when pressing DOWN key on last tab", () => {
    const wrapper = mockWrapperWithSelectedId("baz");

    const divEl = wrapper.find("div");
    divEl.simulate("keydown", { keyCode: DOWN_ARROW_KEY_CODE });

    expect(wrapper.state()).toMatchSnapshot();
  });

  it("should update correctly when pressing DOWN key when not on last tab", () => {
    const wrapper = mockWrapperWithSelectedId("bar");

    const divEl = wrapper.find("div");
    divEl.simulate("keydown", { keyCode: DOWN_ARROW_KEY_CODE });

    expect(wrapper.state()).toMatchSnapshot();
  });

  it("should update correctly when not pressing on arrow key", () => {
    const wrapper = mockWrapperWithSelectedId("bar");

    const divEl = wrapper.find("div");
    divEl.simulate("keydown", { keyCode: 1234 });

    expect(wrapper.state()).toMatchSnapshot();
  });

  it("should focus on tab", () => {
    const wrapper = mount(
      <Tabs {...defaultProps}>
        <h1>Hello World</h1>
      </Tabs>
    );

    const cb = jest.fn();
    const { onTabMount } = wrapper.instance().getChildContext();
    onTabMount("foo");
    onTabMount("bar");
    onTabMount("baz", cb);
    wrapper.setState({ selectedId: "bar" });

    const divEl = wrapper.find("div");
    divEl.simulate("keydown", { keyCode: DOWN_ARROW_KEY_CODE });
    expect(cb).toHaveBeenCalledTimes(1);
  });

  it("should call custom onKeyDown callback", () => {
    const cb = jest.fn();
    const wrapper = mount(
      <Tabs {...defaultProps} onKeyDown={cb}>
        <h1>Hello World</h1>
      </Tabs>
    );

    const divEl = wrapper.find("div");
    divEl.simulate("keydown", { keyCode: DOWN_ARROW_KEY_CODE });

    expect(cb).toHaveBeenCalledTimes(1);
  });
});

describe("onSelect", () => {
  it("should call onSelect function when activating a new tab", () => {
    const cb = jest.fn();
    const wrapper = mount(
      <Tabs {...defaultProps} onSelect={cb}>
        <h1>Hello World</h1>
      </Tabs>
    );

    const { onActivate } = wrapper.instance().getChildContext();
    onActivate("foo");

    expect(cb).toHaveBeenCalledTimes(1);
    expect(cb.mock.calls[0][0]).toEqual("foo");
  });

  it("should call onSelect function when pressing LEFT/UP arrows", () => {
    const cb = jest.fn();
    const wrapper = mount(
      <Tabs {...defaultProps} onSelect={cb}>
        <h1>Hello World</h1>
      </Tabs>
    );

    const { onTabMount } = wrapper.instance().getChildContext();
    onTabMount("foo");
    onTabMount("bar");
    onTabMount("baz");
    wrapper.setState({ selectedId: "foo" });

    const divEl = wrapper.find("div");
    divEl.simulate("keydown", { keyCode: LEFT_ARROW_KEY_CODE });
    divEl.simulate("keydown", { keyCode: UP_ARROW_KEY_CODE });

    expect(cb).toHaveBeenCalledTimes(2);
    expect(cb.mock.calls).toMatchSnapshot();
  });

  it("should call onSelect function when pressing RIGHT/DOWN arrows", () => {
    const cb = jest.fn();
    const wrapper = mount(
      <Tabs {...defaultProps} onSelect={cb}>
        <h1>Hello World</h1>
      </Tabs>
    );

    const { onTabMount } = wrapper.instance().getChildContext();
    onTabMount("foo");
    onTabMount("bar");
    onTabMount("baz");
    wrapper.setState({ selectedId: "foo" });

    const divEl = wrapper.find("div");
    divEl.simulate("keydown", { keyCode: RIGHT_ARROW_KEY_CODE });
    divEl.simulate("keydown", { keyCode: DOWN_ARROW_KEY_CODE });

    expect(cb).toHaveBeenCalledTimes(2);
    expect(cb.mock.calls).toMatchSnapshot();
  });

  it("should not error when onSelect function is undefined", () => {
    const wrapper = mount(
      <Tabs {...defaultProps} onSelect={undefined}>
        <h1>Hello World</h1>
      </Tabs>
    );

    const { onActivate } = wrapper.instance().getChildContext();
    expect(() => onActivate("foo")).not.toThrow();
  });
});

describe("When Tabs is controlled", () => {
  it("should not change selectedId when a Tab is activated", () => {
    const wrapper = mount(
      <Tabs {...defaultProps} selectedId="foo">
        <h1>Hello World</h1>
      </Tabs>
    );

    const { onActivate, selectedId } = wrapper.instance().getChildContext();
    onActivate("bar");

    expect(selectedId).toEqual("foo");
  });

  it("should not change selectedId when pressing an arrow key", () => {
    const cb = jest.fn();
    const wrapper = mount(
      <Tabs {...defaultProps} selectedId="bar" onSelect={cb}>
        <h1>Hello World</h1>
      </Tabs>
    );

    const { onTabMount } = wrapper.instance().getChildContext();
    onTabMount("foo");
    onTabMount("bar");
    onTabMount("baz");

    const divEl = wrapper.find("div");
    divEl.simulate("keydown", { keyCode: RIGHT_ARROW_KEY_CODE });

    expect(wrapper.instance().getChildContext().selectedId).toEqual("bar");
    expect(cb).toHaveBeenCalledTimes(1);
    expect(cb.mock.calls[0][0]).toEqual("baz");
  });
});
