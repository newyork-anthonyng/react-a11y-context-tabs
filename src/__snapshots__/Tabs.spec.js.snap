// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Context should append tab index only if not present 1`] = `
Array [
  "foo",
  "bar",
]
`;

exports[`Context should expose correct context 1`] = `
Object {
  "forceRenderTabPanel": false,
  "onActivate": [Function],
  "onTabMount": [Function],
  "selectedId": null,
  "selectedTabClassName": "tab--selected",
  "selectedTabPanelClassName": "tabPanel--selected",
}
`;

exports[`Context should should update selectedId with onActivate function 1`] = `
Object {
  "selectedId": "foo",
}
`;

exports[`Key down events should update correctly when not pressing on arrow key 1`] = `
Object {
  "selectedId": "bar",
}
`;

exports[`Key down events should update correctly when pressing DOWN key on last tab 1`] = `
Object {
  "selectedId": "foo",
}
`;

exports[`Key down events should update correctly when pressing DOWN key when not on last tab 1`] = `
Object {
  "selectedId": "baz",
}
`;

exports[`Key down events should update correctly when pressing LEFT key on first tab 1`] = `
Object {
  "selectedId": "baz",
}
`;

exports[`Key down events should update correctly when pressing LEFT key when not on first tab 1`] = `
Object {
  "selectedId": "foo",
}
`;

exports[`Key down events should update correctly when pressing RIGHT key on last tab 1`] = `
Object {
  "selectedId": "foo",
}
`;

exports[`Key down events should update correctly when pressing RIGHT key when not on last tab 1`] = `
Object {
  "selectedId": "baz",
}
`;

exports[`Key down events should update correctly when pressing UP key on first tab 1`] = `
Object {
  "selectedId": "baz",
}
`;

exports[`Key down events should update correctly when pressing UP key when not on first tab 1`] = `
Object {
  "selectedId": "foo",
}
`;

exports[`onSelect should call onSelect function when pressing LEFT/UP arrows 1`] = `
Array [
  Array [
    "baz",
  ],
  Array [
    "bar",
  ],
]
`;

exports[`onSelect should call onSelect function when pressing RIGHT/DOWN arrows 1`] = `
Array [
  Array [
    "bar",
  ],
  Array [
    "baz",
  ],
]
`;
